### PR TITLE
lzbt: abstraction for multiple backends

### PR DESCRIPTION
**Pull request by RaitoBezarius**
This generates `lzbt-systemd` binary instead of `lzbt` which is using a special systemd-specific entrypoint.

This is part of the effort to enable multiple backends.

### CODE REVIEW

File: rust/tool/src/esp.rs [hunk 1]
```diff
@@ -1,65 +1,29 @@
...
+// The const generics is necessary because:
+// (a) we cannot return Trait objects in trait definitions, see https://rust-lang.github.io/impl-trait-initiative/explainer/rpit_trait.html
+// (b) we cannot return std::slice::Iter<'_, PathBuf> because PathBuf is owned
+// (c) we cannot return std::slice::Iter<'_, &PathBuf> because then Item = &&PathBuf
+// (d) we cannot return std::slice::Iter<'_, Path> because Path is unsized
+// (e) we cannot return std::slice::Iter<'_, &Path> because then Item = &&Path
+// (f) we can bring an associated type `BorrowedIterator` and then because we want to return
+// borrows, Rust needs a lifetime and the struct cannot hold the borrow so you need to bring a
+// PhantomData<'a ()> inside the structure to keep the lifetime, this is needlessly complicated
+// when this solution ensures proper monomorphisation at a small complexity cost.
```
Comment (nikstur) [inline rust/tool/src/esp.rs:17]: I don't think we need this in the source code. It will not age well.

File: rust/tool/src/systemd/cli.rs
Comment (nikstur) [inline rust/tool/src/systemd/cli.rs:1]: This whole `systemd/` tree is not supposed here, right?

File: rust/tool/Cargo.toml [hunk 2]
```diff
@@ -3,17 +3,11 @@ name = "lanzaboote_tool"
...
-path = "src/main.rs"
-
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
 anyhow = "1.0.71"
-clap = { version = "4.3.1", features = ["derive"] }
 goblin = "0.6.1"
-nix = { version = "0.26.2", default-features = false, features = [ "fs" ] }
 serde = { version = "1.0.163", features = ["derive"] }
```
Comment (nikstur) [inline rust/tool/Cargo.toml:11]: If we turn this into a library we should make the constraints less concrete.

```suggestion
serde = { version = "1", features = ["derive"] }
```

**Review by nikstur — COMMENTED** [review]
I'd go for a different directory structure: 

`tool/`:
- `lib/`
- `systemd-boot/

`tool` should be a Rust workspace.

File: rust/tool/src/esp.rs (hunk 1 above)
Comment (RaitoBezarius) [reply rust/tool/src/esp.rs:17]: I wanted to write down this for myself mostly but yeah

**Comment by nikstur** [comment]
How would we implement generic functionality in this scheme (e.g. key rotation)?

**Comment by RaitoBezarius** [comment]
> How would we implement generic functionality in this scheme (e.g. key rotation)?

You'd tuck in the shared crate IMHO?

File: rust/tool/systemd/Cargo.lock
Comment (nikstur) [inline rust/tool/systemd/Cargo.lock:1]: I think this shouldn't have its own lock but use the one at the workspace root.

File: rust/tool/systemd/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,27 @@
+[package]
+name = "lzbt-systemd"
+version = "0.3.0"
```
Comment (nikstur) [inline rust/tool/systemd/Cargo.toml]:
```suggestion
version.workspace = true
```

File: rust/tool/systemd/src/main.rs [hunk 4]
```diff
@@ -1,16 +1,9 @@
...
 mod install;
-mod os_release;
-mod pe;
-mod signature;
-mod systemd;
-mod utils;
+mod version;
 
 use clap::Parser;
-
```
Comment (nikstur) [inline rust/tool/systemd/src/main.rs:13]: nit: please keep this line it separates external dependencies from internal dependencies.

**Review by nikstur — APPROVED** [review]
Just fixed all my nits myself. This is good to go now.

**Pull request merged**